File: src/test/java/no/nav/syfo/testhelper/OidcTestHelper.java
package no.nav.syfo.testhelper;

import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import no.nav.security.oidc.context.*;
import no.nav.security.spring.oidc.test.JwtTokenGenerator;
import no.nav.syfo.oidc.OIDCIssuer;

import java.text.ParseException;

public class OidcTestHelper {

    public static void loggInnVeilederAzure(OIDCRequestContextHolder oidcRequestContextHolder, String veilederIdent) throws ParseException {
        JWTClaimsSet claimsSet = JWTClaimsSet.parse("{\"NAVident\":\"" + veilederIdent + "\"}");
        SignedJWT jwt = JwtTokenGenerator.createSignedJWT(claimsSet);
        settOIDCValidationContext(oidcRequestContextHolder, jwt, OIDCIssuer.AZURE);
    }

    public static void loggInnVeileder(OIDCRequestContextHolder oidcRequestContextHolder, String subject) {
        SignedJWT jwt = JwtTokenGenerator.createSignedJWT(subject);

        settOIDCValidationContext(oidcRequestContextHolder, jwt, OIDCIssuer.INTERN);
    }

    public static void loggInnBruker(OIDCRequestContextHolder oidcRequestContextHolder, String subject) {
        SignedJWT jwt = JwtTokenGenerator.createSignedJWT(subject);
        settOIDCValidationContext(oidcRequestContextHolder, jwt, OIDCIssuer.EKSTERN);
    }

    private static void settOIDCValidationContext(OIDCRequestContextHolder oidcRequestContextHolder, SignedJWT jwt, String issuer) {
        TokenContext tokenContext = new TokenContext(issuer, jwt.serialize());
        OIDCClaims oidcClaims = new OIDCClaims(jwt);
        OIDCValidationContext oidcValidationContext = new OIDCValidationContext();
        oidcValidationContext.addValidatedToken(issuer, tokenContext, oidcClaims);
        oidcRequestContextHolder.setOIDCValidationContext(oidcValidationContext);
    }

    public static OIDCValidationContext lagOIDCValidationContextIntern(String subject) {
        return lagOIDCValidationContext(subject, OIDCIssuer.INTERN);
    }

    public static OIDCValidationContext lagOIDCValidationContextEkstern(String subject) {
        return lagOIDCValidationContext(subject, OIDCIssuer.EKSTERN);
    }

    private static OIDCValidationContext lagOIDCValidationContext(String subject, String issuer) {
        SignedJWT jwt = JwtTokenGenerator.createSignedJWT(subject);
        TokenContext tokenContext = new TokenContext(issuer, jwt.serialize());
        OIDCClaims oidcClaims = new OIDCClaims(jwt);
        OIDCValidationContext oidcValidationContext = new OIDCValidationContext();
        oidcValidationContext.addValidatedToken(issuer, tokenContext, oidcClaims);
        return oidcValidationContext;
    }

    public static void loggUtAlle(OIDCRequestContextHolder oidcRequestContextHolder) {
        oidcRequestContextHolder.setOIDCValidationContext(null);
    }

}

File: src/test/kotlin/no/nav/syfo/testhelper/OidcTestHelper.kt
package no.nav.syfo.testhelper

import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.SignedJWT
import io.mockk.every
import no.nav.security.oidc.context.*
import no.nav.security.oidc.test.support.JwtTokenGenerator
import no.nav.syfo.api.auth.OIDCIssuer
import no.nav.syfo.api.auth.OIDCUtil
import java.text.ParseException

object OidcTestHelper {
    @JvmStatic
    @Throws(ParseException::class)
    fun loggInnVeilederAzureV2(
        oidcRequestContextHolder: OIDCRequestContextHolder,
        veilederIdent: String
    ) {
        val claimsSet = JWTClaimsSet.parse("{\"NAVident\":\"$veilederIdent\", \"azp\":\"clientid\"}")
        val jwt = JwtTokenGenerator.createSignedJWT(claimsSet)
        settOIDCValidationContext(oidcRequestContextHolder, jwt, OIDCIssuer.VEILEDER_AZURE_V2)
    }

    private fun settOIDCValidationContext(
        oidcRequestContextHolder: OIDCRequestContextHolder,
        jwt: SignedJWT,
        issuer: String
    ) {
        val tokenContext = TokenContext(issuer, jwt.serialize())
        val oidcClaims = OIDCClaims(jwt)
        val oidcValidationContext = OIDCValidationContext()
        oidcValidationContext.addValidatedToken(issuer, tokenContext, oidcClaims)
        oidcRequestContextHolder.oidcValidationContext = oidcValidationContext
    }

    @JvmStatic
    fun loggUtAlle(oidcRequestContextHolder: OIDCRequestContextHolder) {
        oidcRequestContextHolder.oidcValidationContext = null
    }

    fun mockOIDCUtils(contextHolder: OIDCRequestContextHolder) {
        every {
            OIDCUtil.tokenFraOIDC(contextHolder, OIDCIssuer.VEILEDER_AZURE_V2)
        } returns "token"

        every {
            OIDCUtil.getSubjectInternAzureV2(contextHolder)
        } returns UserConstants.VEILEDER_ID

        every {
            OIDCUtil.getAzpAzureV2(contextHolder)
        } returns "azp"
    }
}
